dll/win32/kernel32/client/proc.c

```diff
@@ -486,6 +486,13 @@ BaseProcessStartup(PPROCESS_START_ROUTINE lpStartAddress)
                                &lpStartAddress,
                                sizeof(PPROCESS_START_ROUTINE));
 
+	#if defined(__i386__) 
+	    /*Set EBX = PEB */
+	    asm volatile (".intel_syntax noprefix");
+	    asm volatile ("mov ebx, fs:0x30");
```

Comment:
Do you have documentation for this? We'll probably need a test showing that this is correct.

Reply:
I will look for official documentation, however you can run the next code with Flat assembler. All windows NT based versions will result in YES, all windows 9x will return in No.

format PE GUI
entry Start

include 'win32a.inc'

section '.text' executable code data readable writeable

Start:
mov eax,dword [fs:0x30] ;Get PEB from TEB
cmp ebx,eax ;compare EBX with fs:0x30
jne ThisIsNotCorrect ;jump if not equal

push 0
push 0
push yes
push 0
call [MessageBoxA]
jmp return

ThisIsNotCorrect:
push 0
push 0
push no
push 0
call [MessageBoxA]

return:
ret

yes db 'EBX is set to PEB (fs:0x30)',0
no db 'EBX is NOT set to PEB (fs:0x30)',0

section '.idata' import data readable

  library user,'USER32.DLL'

  import user,\
         MessageBoxA,'MessageBoxA'

Reply:
Here is a reference, see chapter "II.  Application  start" from http://www.fysnet.net/docs/apj_2004_144.pdf

Reply:
> Added register EBX is set to PEB at new thread startup just like all Windows NT+ versions do

That sounds like something that must be done in the RTL or elsewhere.

Reply:
The reason why I set EBX just before the startAddress is called is because it cannot be changed somewhere else in the code. For instance, the function BaseProcessStartup can change EBX to any value any time. If there is more code between changing EBX and calling startAddress, how bigger the possibility is that the compiler changes EBX somewhere.

Reply:
Does it also work on 32-bit versions of Windows 7, 10... ?

Reply:
Note that in my previous comment I mentioned places where registers can be set (either via direct assembly, or using portable C code using the Context structure) without being modified afterwards.

Reply:
@HBelusca Yes I checked it on all the Windows NT+ versions. even 7 and 10. Some of my programs use it to detect if it runs on w9x or wNT. That's why I noticed the difference on ReactOS. My program was not functioning because it was expecting a w9x system. It can still be a coincidence, however, it is quite stable coincidence through all the Windows NT-kernel systems.

Reply:
Do you mean the SetThreadContext() function?

Reply:
Hmm this looks okay though...

```
_BaseProcessStartup@4:
  00001DE2: 55                 push        ebp
  00001DE3: 89 E5              mov         ebp,esp
  00001DE5: 57                 push        edi
  00001DE6: 56                 push        esi
  00001DE7: 53                 push        ebx
  00001DE8: 83 EC 3C           sub         esp,3Ch
  00001DEB: EB 32              jmp         00001E1F
  00001DED: C7 44 24 0C 04 00  mov         dword ptr [esp+0Ch],4
            00 00
  00001DF5: 8D 45 08           lea         eax,[ebp+8]
  00001DF8: 89 44 24 08        mov         dword ptr [esp+8],eax
  00001DFC: C7 44 24 04 09 00  mov         dword ptr [esp+4],9
            00 00
  00001E04: C7 04 24 FE FF FF  mov         dword ptr [esp],0FFFFFFFEh
            FF
  00001E0B: FF 15 00 00 00 00  call        dword ptr [__imp__NtSetInformationThread@16]
  00001E11: 83 EC 10           sub         esp,10h
  00001E14: FF 55 08           call        dword ptr [ebp+8]
  00001E17: 89 04 24           mov         dword ptr [esp],eax
  00001E1A: E8 00 00 00 00     call        _ExitThread@4
  00001E1F: 90                 nop
  00001E20: 8D 45 CC           lea         eax,[ebp-34h]
  00001E23: 8D 15 84 0E 00 00  lea         edx,[.rdata+0E84h]
  00001E29: E8 00 00 00 00     call        __SEH3$_RegisterFrame
  00001E2E: 66 90              xchg        ax,ax
  00001E30: EB BB              jmp         00001DED
  00001E32: 80 3D 00 00 00 00  cmp         byte ptr [_BaseRunningInServerProcess],0
            00
  00001E39: 75 0B              jne         00001E46
  00001E3B: 8B 45 E0           mov         eax,dword ptr [ebp-20h]
  00001E3E: 89 04 24           mov         dword ptr [esp],eax
  00001E41: E8 87 FE FF FF     call        _ExitProcess@4
  00001E46: 8B 45 E0           mov         eax,dword ptr [ebp-20h]
  00001E49: 89 04 24           mov         dword ptr [esp],eax
  00001E4C: E8 00 00 00 00     call        _ExitThread@4
  00001E51: EB FE              jmp         00001E51
```

Reply:
@HBelusca 
Sorry I misunderstood your comment. I think I understand now. 

Currently BaseInitializeContext sets ebx to the parameter for function startAddress. BaseProcessStartThunk (kernel32/client/i386/thread.S) will push EBX on stack and calls BaseProcessStartup which starts startAddress.

It is possible to change the register for the parameter (initiated by BaseInitializeContext) to ECX and set EBX to PEB, I should try that. However, it is risky because the function BaseProcessStartup() can still change EBX (now or in the future) because the C-code is compiled in that order.

Reply:
@ThFabba was writing my comment before I read yours :) The assembly looks OK! Do you want to risk it and change thread.C, with the possibility that in the future the C-code/compiler changes EBX?

Reply:
I've another question: so far you've tested your code with a GCC build of ROS? If so, can you test an MSVC build of ROS instead (available through www.reactos.org/getbuilds as usual)? I'm curious to know whether it could be a compiler effect (and we also know that MS windows is built with MSVC).

Reply:
Well I'd like to first find out why it's broken right now. Because from looking at the theory, ebx should already point to the PEB as-is.

Reply:
@ThFabba Currently EBX is set to the parameter (for function startAddress.). This is done by BaseInitializeContext.  BaseProcessStartThunk and BaseProcessStartup do not change EBX.

@HBelusca Will do, however I first need to setup the build environment on my Windows system. This may take a while, iĺl do it tomorrow :)

Reply:
The parameter _is_ the PEB. See https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/kernel32/client/proc.c;hb=fc9bc9390d1b6548acc21922546711042483e5b8#l4224

Reply:
@ThFabba I tried to create an C example, however, my C compiler sets EBX to 1 before it's executing my code. So please consider this assembler example that can be compiled with FASM : https://pastebin.com/74mJx5md . 

This example converts  register EBX, [fs:0x30] and value NtCurrentTeb()+0x30 to a readable hex-string. These should all result in the same values because both are address to the PEB. Please see screen-shots for results:

Clean Reactos install: https://imgur.com/k3IVdBn
Windows XP (32 bit): https://imgur.com/JySs2xo
Windows 7(64 bit): https://imgur.com/qqe1uaD
Windows 10 (64 bit): https://imgur.com/GLGl4zb  (no full desktop screenshot because itś my home system :))

As you can see the pointers differ in ReactOS. They do not differ in Windows. Is there a reason for that? In ReactOs with my changes you get the next screenshot, wih the same results as the windows systems:

ReactOS (with my changes): https://imgur.com/ied1CDX

Reply:
The real question now is whether, when you use MSVC builds of ReactOS (e.g. https://iso.reactos.org/bootcd/reactos-bootcd-0.4.9-dev-690-gff9c737-x86-msvc-win-dbg.7z ) the problem still happens.

Reply:
Trying it, and got same results, the pointers EBX and the real PEB differ